File: built-in plugins dsp/mb-sampler-drums.cpp
/**
 * MB Drum Sampler
 * Category : instrument
 * Type     : sampler
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Multi-layer drum samples
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_SAMPLER_DRUMS_H
#define MB_SAMPLER_DRUMS_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbSamplerDrums : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-sampler-drums";
    static constexpr const char* PLUGIN_NAME    = "MB Drum Sampler";
    static constexpr const char* PLUGIN_TYPE    = "sampler";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float kit = 1f;  // range [1, 8]
    float punch = 0.7f;  // range [0, 1]
    float volume = 0.85f;  // range [0, 1]
    };

    MbSamplerDrums() = default;
    ~MbSamplerDrums() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.kit = std::clamp(params.kit, 1f, 8f);
        params.punch = std::clamp(params.punch, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Drum Sampler
        return input;
    }
};

#endif // MB_SAMPLER_DRUMS_H

File: built-in plugins dsp/mb-sampler-drums.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbSamplerDrumsPlugin: PluginDefinition = { id: 'mb-sampler-drums', slug: 'mb-sampler-drums', name: 'MB Drum Sampler', category: 'instrument', type: 'sampler', version: '1.0.0', description: 'Multi-layer drum samples', author: 'Max Booster', oscillators: [], envelope: { attack: 0.001, decay: 0.3, sustain: 0, release: 0.2 }, parameters: [{ id: 'kit', name: 'Kit', type: 'float', defaultValue: 1, minValue: 1, maxValue: 8, automatable: false }, { id: 'punch', name: 'Punch', type: 'float', defaultValue: 0.7, minValue: 0, maxValue: 1, automatable: true }, { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.85, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { kit: 1, punch: 0.7, volume: 0.85 } };

export default MbSamplerDrumsPlugin;
